### ISSUE TITLE
[BUG] Some 3006.6 minions not compatible with 3006.7 master

### ISSUE THREAD

**Issue by MartinEmrich** [saltstack/salt#66126]
**Description**
After some new VMs this week got 3006.7 from the repository out-of-the-box, which did not work with Salt master 3006.6, I had to upgrade the salt-master to 3006.7. But now some 3006.6 minions are no longer reachable.

They log every 10 seconds:
````
2024-02-22 16:50:10,665 [salt.crypt       :823 ][ERROR   ][37572] The Salt Master has rejected this minion's public key.
To repair this issue, delete the public key for this minion on the Salt Master.
The Salt Minion will attempt to re-authenicate.
````
(sic)

Restarting the minion did not help.

So neither updating master first nor minion first resulted in a stable upgrade experience.

After upgrading the minion to 3006.7, reloading systemd and restarting the minion, the connection works.

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [x] on-prem machine
- [x] VM (Virtualbox, KVM, etc. please specify)
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [ ] classic packaging
- [x] onedir packaging
- [x] used bootstrap to install

**Steps to Reproduce the behavior**
(Include debug logs if possible and relevant)

**Expected behavior**
I would hope for minion-master compatibility within any minor-version mix, and at least one step difference in major version.
(Of course with only the features both support).

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

Minion:
```yaml
Salt Version:
          Salt: 3006.6

Python Version:
        Python: 3.10.13 (main, Nov 15 2023, 04:34:27) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.3
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.19.1
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.14.2
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: centos 7.9.2009 Core
        locale: utf-8
       machine: x86_64
       release: 3.10.0-1160.108.1.el7.x86_64
        system: Linux
       version: CentOS Linux 7.9.2009 Core```
```
</details>

Master:
<details>
```yaml
Salt Version:
          Salt: 3006.7

Python Version:
        Python: 3.10.13 (main, Feb 19 2024, 03:31:20) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.3
       libgit2: 1.3.0
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.19.1
        pygit2: 1.7.0
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.15.1
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: centos 7.9.2009 Core
        locale: utf-8
       machine: x86_64
       release: 3.10.0-1160.105.1.el7.x86_64
        system: Linux
       version: CentOS Linux 7.9.2009 Core

```
</details>

**Comment by MartinEmrich** [comment]
I grabbed an affected minion, and did a diff between "before" and "after" installing the upgrade, and browsed through the code changes.

Most seems to be 3rd-party updates (OpenSSL), but I noticed code relating to #52289

I just skimmed through it, but it seems to be about Line endings on the minion/master keys.

Indeed: The affected minions have newline (0x0A) at the end of `/etc/salt/pki/minion/minion.pub` and `minion.pem`.
Other minions have no newline at the end of the keypair, they still work with minion 3006.6 and a 3006.7 master.

I have no Idea though what causes the difference, both the reachable and the dead minions range from many years old to quite new.

**Comment by darkpixel** [comment]
@MartinEmrich If you edit your minion.pub and minion.pem using ```vi -b minion.pub``` and then run ```:set noeol``` and then ```:wq``` and start the salt-minion service again, it will come back online until this gets fixed.

**Comment by robertsonjaNIH** [comment]
Thank you @MartinEmrich for opening this, we were bitten by the same issue today when upgrading two syndics supporting a compute cluster (bare metal). We had upgraded our upstream master already but only took it to v3006.6, so we didn't see the problem come up then.

Thank you also @darkpixel for the fix, that resolved the issue for us! We were able to clean up all of our minions containing newlines in their minion.pub's and minion.pem's and restart the salt-minion service, which cleared the problem for the affected hosts.

We dug a little deeper and have some info which may be useful in hunting down the root cause:

- It turns out our syndics themselves had newline characters in their respective minion.pem and minion.pub files as well. Interestingly however, their syndic_master.pub files did _not_ contain newline characters, although the creation time of syndic_master.pub and minion.pub/minion.pem was less than 30 minutes apart
- the version of Salt installed which performed the key creations was salt-3004.2-1.el8.noarch (both syndics are CentOS Stream 8 systems). The version of OpenSSL installed at that time was openssl-1:1.1.1k-6.el8.x86_64
- we don't believe the issue is strictly tied to a specific OS, as we saw impacted minions running a range of OS's, including but not limited to: RHEL7, RHEL8, CentOS Linux 7, CentOS Stream 8, and openSUSE Leap 15

Hopefully something in there is useful!

**Comment by darkpixel** [comment]
Paging @twangboy to the white courtesy phone... :)

**Comment by dwoz** [comment]
@darkpixel @twangboy is a away for a few days. I will take a look at this soon.

**Comment by darkpixel** [comment]
Hopefully soaking up some sun somewhere fun.

**Comment by gandihar** [comment]
I've been bitten by the same bug. It has knocked every one of my Minions offline. Every one not running 3006.7 will not come back. I've tried darkpixel's fix on a couple, but the Master rejects them anyway. I've compared sha256sum's across systems and they do match for the master.pub and minion.pub|$minion_id.pub files, but they continually end up in the Denied list until I:

1. Shutdown the minion daemon.
2. Nuke the minion key on both the Minion and Master.
3. Upgrade the Minion package to 3006.7
4. Start the salt-minion service
5. Pray I don't see the minion listed under both Denied and Unaccepted keys at the same  time.

I have some legacy Minions that cannot upgrade to 3006.7, so I would like to help get this fixed if I can.

**Comment by darkpixel** [comment]
> I've been bitten by the same bug. It has knocked every one of my Minions offline. Every one not running 3006.7 will not come back. I've tried darkpixel's fix on a couple, but the Master rejects them anyway. I've compared sha256sum's across systems and they do match for the master.pub and minion.pub|$minion_id.pub files, but they continually end up in the Denied list until I:
> 
> 1. Shutdown the minion daemon.
> 2. Nuke the minion key on both the Minion and Master.
> 3. Upgrade the Minion package to 3006.7
> 4. Start the salt-minion service
> 5. Pray I don't see the minion listed under both Denied and Unaccepted keys at the same  time.
> 
> I have some legacy Minions that cannot upgrade to 3006.7, so I would like to help get this fixed if I can.

It's definitely the line at the end of the file.
```service salt-minion stop```
```vi -b /etc/salt/pki/minion/*```
```:set noeol```
```:wn``` until it complains you've reached the last file.
```:wq```
```salt-call test.ping```
If it's accepted, great.  If not, go to the master and ```salt-key -d name-of-minion``` and try ```salt-call test.ping``` again.
It should work, then do ```service salt-minion restart````

**Comment by jzandbergen** [comment]
The problem is in some crypto stuff. Public keys with a end-of-line character are not valid anymore on the salt master. To fix this:

1. stop the salt-minion
2. delete the keys on the salt master
3. `truncate -s -1 /etc/salt/pki/minion/minion.pub` (this removes the end of line character.
4. start the salt-minion
5. accept the key

**Comment by Z10N0110** [comment]
Due to this bug, I think for production usage should always do the version locking to specific minor version in the repository file like I mentioned in #66121

**Comment by MartinEmrich** [comment]
@darkpixel thanks for the hint, but as I have to SSH into every host now anyways, I will just upgrade the minions to 3006.7. For the few I tried so far, that fixed it.

**Comment by jzandbergen** [comment]
** warning: this is a quick workaround to get it working again. Use at your own peril **

To get stuff working again:

in `<salt dir>/salt/crypt.py` line 87 change
```
return "\n".join(key.strip().splitlines())
```
in
```
return key
```

Then:
1. stop the salt-master
2. delete the keys in `/etc/salt/pki/master/minions_denied`
3. start the salt-master

The minions should now be available again.

**Comment by kiniou** [comment]
Adding `salt.crypt.clean_key` on the right side of [this condition](https://github.com/saltstack/salt/pull/65989/files#diff-54eee5af67c903f2c13cda310d921229b74b3cfb7b08fa5886a32fba63ad778cL378-R374) worked for me.

```python
if salt.crypt.clean_key(pubfn_handle.read()) != salt.crypt.clean_key(load["pub"])
```

**Comment by dwoz** [comment]
> Adding `salt.crypt.clean_key` on the right side of [this condition](https://github.com/saltstack/salt/pull/65989/files#diff-54eee5af67c903f2c13cda310d921229b74b3cfb7b08fa5886a32fba63ad778cL378-R374) worked for me.
> 
> ```python
> if salt.crypt.clean_key(pubfn_handle.read()) != salt.crypt.clean_key(load["pub"])
> ```

This is likely the bast path forward for anyone looking for a quick fix that doesn't require changes on the minion.

**Comment by jzandbergen** [comment]
> Adding `salt.crypt.clean_key` on the right side of [this condition](https://github.com/saltstack/salt/pull/65989/files#diff-54eee5af67c903f2c13cda310d921229b74b3cfb7b08fa5886a32fba63ad778cL378-R374) worked for me.
> 
> ```python
> if salt.crypt.clean_key(pubfn_handle.read()) != salt.crypt.clean_key(load["pub"])
> ```

Thank you for your feedback @kiniou i've created a draft PR on your solution. In case the code is not shutdown i'll credit you where it is appropriate (I'll have to find out). Also, please let me know if I'm breaking convention by using your solution in my PR. I'll retract it if that is the case.

**Comment by kiniou** [comment]
>In case the code is not shutdown i'll credit you where it is appropriate (I'll have to find out). 
Don't worry about this too much, I'm not that really attached to the solution 😉

>Also, please let me know if I'm breaking convention by using your solution in my PR.
I don't think you'll break any conventions since the clean_key was already done on incoming minion's key before (you can walk through commits history to see that).

**Comment by rasgard** [comment]
quick and dirty 

minions:
sudo truncate -s -1 /etc/salt/pki/minion/minion.pem && sudo truncate -s -1 /etc/salt/pki/minion/minion.pub && sudo systemctl restart salt-minion
master: 
sudo salt-key -A --include-denied

if minion fails check pem pub for missing "-" end of last line

**Comment by lomeroe** [comment]
Should the public key from the master be run through the "clean_key" function too (i.e. payload["pub_key"] on line 1177 of crypt.py)?  The file read of the file system is, but it doesn't look like the public key coming from the master is....

https://github.com/saltstack/salt/blob/ce9d09bc9c8f48d588529177023b19f11f47ef1a/salt/crypt.py#L1174-L1177

I had to modify L1177 on a 3006.7 minion for it to validate the master's public key (after manually applying the fixes from #66153).  Granted, the master was not yet at 3006.7, so maybe getting the master to 3006.7 would also take care of it, but I was in a bit of a chicken/egg scenario upgrading a MoM to 3006.7 before moving to the syndics (the MoM's minion is connected to the syndic masters).

```
--- /tmp/crypt.3007     2024-03-08 15:10:49.775314610 -0600
+++ /opt/saltstack/salt/lib/python3.10/site-packages/salt/crypt.py      2024-03-08 14:47:49.329121607 -0600
@@ -1174,7 +1174,7 @@
             with salt.utils.files.fopen(m_pub_fn) as fp_:
                 local_master_pub = clean_key(fp_.read())

-            if payload["pub_key"] != local_master_pub:
+            if clean_key(payload["pub_key"]) != local_master_pub:
                 if not self.check_auth_deps(payload):
                     return ""

```

**Comment by twangboy** [comment]
This has been fixed here for 3006.x: https://github.com/saltstack/salt/pull/66153
And here for 3007.x: https://github.com/saltstack/salt/pull/66161
